Feature/user admin

Syapa723 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #105 
- 작업 요약: 회원관리페이지 회원권한부여 기능 개발

## 📄 상세 내용
- [x] 관리자페이지에서 회원에게 권한(admin, staffm general)을 선택해서 부여하는 기능
- [x] 관리자페이지에 접근 권한 검사를 위한 permissions.py 작성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Syapa723 committed:
✨ feat: 회원관리 권한부여 기능 개발

회원관리 페이지(user admin page)에서 회원별 관리자 권한을 부여기능 개발

Umdoong commented:
아이고... 테스트가 79%가 나와서 터졌네요 ㅠ
main으로 pr 날리셨어용 ㅜ

Syapa723 committed:
✅ test: admin_test -> test_admin 파일 이름 변경

테스트 커버리지 상향을 위한 admin_test 파일 이름 변경
admin_test -> test_admin

Umdoong commented on apps/users/serializers/admin_serializers.py:
```diff
@@ -0,0 +1,82 @@
+        model = User
+        fields = ["is_staff", "is_superuser"]
+
+    def save(self, **kwargs: Any) -> User:
```
1. service layer에 있는 비즈니스 로직과 겹치는 것 같아용
2. 지금 작성하신 코드에서 serializer.save를 사용하고 있지 않기 때문에 없어도 되는 코드라고 생각되네용

Umdoong commented on apps/users/serializers/admin_serializers.py:
```diff
@@ -0,0 +1,82 @@
+        fields = ["is_staff", "is_superuser"]
+
+    def save(self, **kwargs: Any) -> User:
+        target_user = cast(User, self.context["target_user"])
```
지금 시리얼라이저를 호출하는 곳에서 context를 넘겨주고 있지 않기 때문에 target_user context를 받아올 수 없어용!
context를 밖에서 받아야 사용할 수 있습니다.

SimJunSik commented on apps/users/services/admin_servies.py:
```diff
@@ -0,0 +1,29 @@
+    @staticmethod
+    @transaction.atomic
+    def update_user_permission(user: User, permission: PermissionType) -> User:
+        try:
+            permission_setting = PermissionLevel[permission.upper()].value
+        except KeyError:
+            permission_setting = PermissionLevel.GENERAL.value
```
try-except 를 흐름 제어, 분기처럼 쓰면 좋지 않습니다
KeyError 가 발생하면 logging, raise exception 을 해서 에러를 발생 시키거나

잘못된 permission 이 넘어와도 GENERAL 로 fallback 하는게 스펙이라면 try-except 말고 default dict 를 쓰거나 다른 방법으로 하는게 좋을 것 같네요

Meoyoug replied:
@Syapa723 
준식님이 말씀하신대로 KeyError가 발생한다면 잘못된 권한이 넘어온걸 텐데 Validation에서 잡았어야하는게 아닌가요?

SimJunSik commented on apps/users/services/admin_servies.py:
```diff
@@ -0,0 +1,29 @@
+PermissionType = Literal["admin", "staff", "general"]
+
+
+class PermissionLevel(Enum):
+    ADMIN = (True, True)
+    STAFF = (True, False)
+    GENERAL = (False, False)
```
여기 이 데이터가 정의된 곳만 보면 첫번째 bool 은 무슨 값이고, 두번째 bool 은 무슨 값인지 알 수가 없어서
뒤에 (bool, bool) 부분을 dataclass 로 만들어서 의미를 알 수 있게 개선해보면 좋겠네요

Syapa723 committed:
♻️ refactor: 불필요한 코드 제거 및 피드백 반영 수정

admin_serializers에서 admin_sevices와중복되는 비지니스 로직 코드 삭제
admin_servies bool 데이터 값 명시하도록 수정, try except 구문 .get() 메소드 대체

Syapa723 committed:
✅ test: 수정된 코드에 맞게 test code 수정

수정된 serializer, service에 맞춰서 test code 수정

Syapa723 committed:
♻️ refactor: test code setup data 일부를 setuptestdata로 변경하여 성능 향상

슈퍼유저, 스태프 setuptestdata로 변경

Syapa723 committed:
🎨 style: test code mypy 이슈 해결

test_admin 코드 mypy 에러 수정

Syapa723 committed:
🎨 style: black, isort 코드 포매터 실행

Umdoong commented on apps/users/tests/test_admin.py:
```diff
@@ -0,0 +1,92 @@
+        # self.list_url = reverse("admin_user:users-list")
+
+    # 권한 수정 API 테스트
+    # 슈퍼유저가 일반유저의 권한을 staff으로 변경하는 기능 테스트
```
나중에 슈퍼유저가 staff 유저의 권한을 변경하는 기능 테스트, 어드민이 자기 자신을 변경하려고 시도하는 테스트도 만들어주시면 좋을 것 같습니다.
고생하셨습니다.

Syapa723 committed:
♻️ refactor: serializer, service, view 코드 리팩토링

service 코드와 serializer 코드 병합 후 리팩토링, 수정사항에 맞춰 view 코드 리팩토링

Syapa723 committed:
♻️ refactor: view 코드 리팩토링

user admin view 코드 리팩토링

Syapa723 committed:
💡 chore: 불필요한 import문 제거
